monix-execution/jvm/src/main/scala/monix/execution/schedulers/TracingScheduler.scala

```diff
@@ -0,0 +1,89 @@
+package monix.execution.schedulers
+
+import java.util.concurrent.{ScheduledExecutorService, TimeUnit}
+import monix.execution.misc.Local
+import monix.execution.{Cancelable, Scheduler, UncaughtExceptionReporter, ExecutionModel => ExecModel}
+import scala.concurrent.ExecutionContext
+
+final class TracingScheduler private (
+                                       scheduler: ScheduledExecutorService,
+                                       ec: ExecutionContext,
+                                       r: UncaughtExceptionReporter,
+                                       val executionModel: ExecModel)
+  extends ReferenceScheduler with BatchingScheduler { self =>
+
+  override def executeAsync(r: Runnable): Unit =
+    executeWithTrace(r)
+
+  /** Executes the given task with tracing.
+    *
+    * @param r is the callback to be executed
+    */
+  def executeWithTrace(r: Runnable): Unit = {
+    val oldContext = Local.getContext()
+    ec.execute(new Runnable {
+      override def run = {
+        Local.withContext(oldContext)(r.run())
+      }
+    })
+  }
+
+  /** Schedules a task to run in the future, after `initialDelay`.
+    *
+    * For example the following schedules a message to be printed to
+    * standard output after 5 minutes:
+    * {{{
+    *   val task = scheduler.scheduleOnce(5, TimeUnit.MINUTES, new Runnable {
+    *     def run() = print("Hello, world!")
+    *   })
+    *
+    *   // later if you change your mind ...
+    *   task.cancel()
+    * }}}
+    *
+    * @param initialDelay is the time to wait until the execution happens
+    * @param unit         is the time unit used for `initialDelay`
+    * @param r            is the callback to be executed
+    * @return a `Cancelable` that can be used to cancel the created task
+    *         before execution.
+    */
+  override def scheduleOnce(initialDelay: Long, unit: TimeUnit, r: Runnable): Cancelable = {
+    if (initialDelay <= 0) {
+      executeWithTrace(r)
+      Cancelable.empty
+    } else {
+      val deferred = new Runnable {
+        override def run(): Unit = executeWithTrace(r)
```

Comment:
This anonymous `Runnable` instance isn't OK, but I just realized there's something I need to fix as well, so await further instructions on this one :-)

Reply:
@alexandru understood!